src/main/scala/edu/berkeley/cs/rise/opaque/execution/operators.scala

```diff
@@ -243,14 +248,29 @@ case class EncryptedAggregateExec(
       val (firstRows, lastGroups, lastRows) = childRDD.map { block =>
         val (enclave, eid) = Utils.initEnclave()
         val (firstRow, lastGroup, lastRow) = enclave.NonObliviousAggregateStep1(
-          eid, aggExprSer, block.bytes)
+          eid, aggExprSer, block.bytes, TaskContext.getPartitionId)
         (Block(firstRow), Block(lastGroup), Block(lastRow))
       }.collect.unzip3
 
       // Send first row to previous partition and last group to next partition
-      val shiftedFirstRows = firstRows.drop(1) :+ Utils.emptyBlock
-      val shiftedLastGroups = Utils.emptyBlock +: lastGroups.dropRight(1)
-      val shiftedLastRows = Utils.emptyBlock +: lastRows.dropRight(1)
+      var shiftedFirstRows = Array[Block]()
+      var shiftedLastGroups = Array[Block]()
+      var shiftedLastRows = Array[Block]()
+      if (childRDD.getNumPartitions == 1) {
+        val firstRowDrop = firstRows(0)
+        shiftedFirstRows = firstRows.drop(1) :+ Utils.emptyBlock(firstRowDrop)
```

Comment:
You can just return the `emptyBlock` here?

Reply:
Right -  the result of the `firstRows.drop(1)` call (and the corresponding variables for shiftedLastGroups and shiftedLastRows) is essentially the empty block, but while the `Utils.emptyBlock(block)` function stores nothing in the `EncryptedBlocks` attribute, it does store some LogEntry data. So the block here is not exactly the `emptyBlock`. I coded it like this to make everything easier to read, but can remove the `drop(1)` and `dropRight` calls if you think that's better.